document-store/src/main/java/org/hypertrace/core/documentstore/mongo/query/parser/MongoLetClauseBuilder.java

```diff
@@ -0,0 +1,104 @@
+package org.hypertrace.core.documentstore.mongo.query.parser;
+
+import com.mongodb.BasicDBObject;
+import java.util.Map;
+import org.hypertrace.core.documentstore.expression.impl.AggregateExpression;
+import org.hypertrace.core.documentstore.expression.impl.AliasedIdentifierExpression;
+import org.hypertrace.core.documentstore.expression.impl.ArrayRelationalFilterExpression;
+import org.hypertrace.core.documentstore.expression.impl.ConstantExpression;
+import org.hypertrace.core.documentstore.expression.impl.DocumentArrayFilterExpression;
+import org.hypertrace.core.documentstore.expression.impl.FunctionExpression;
+import org.hypertrace.core.documentstore.expression.impl.IdentifierExpression;
+import org.hypertrace.core.documentstore.expression.impl.KeyExpression;
+import org.hypertrace.core.documentstore.expression.impl.LogicalExpression;
+import org.hypertrace.core.documentstore.expression.impl.RelationalExpression;
+import org.hypertrace.core.documentstore.expression.type.FilterTypeExpression;
+import org.hypertrace.core.documentstore.parser.FilterTypeExpressionVisitor;
+import org.hypertrace.core.documentstore.parser.SelectTypeExpressionVisitor;
+
+// Visitor for building the let clause in lookup stage from filter expressions
+class MongoLetClauseBuilder implements FilterTypeExpressionVisitor {
+
+  private final String subQueryAlias;
+
+  MongoLetClauseBuilder(String subQueryAlias) {
+    this.subQueryAlias = subQueryAlias;
+  }
+
+  @Override
+  public BasicDBObject visit(LogicalExpression expression) {
+    BasicDBObject letClause = new BasicDBObject();
+    for (FilterTypeExpression operand : expression.getOperands()) {
+      letClause.putAll((Map<String, Object>) operand.accept(this));
+    }
+    return letClause;
+  }
+
+  @Override
+  public BasicDBObject visit(RelationalExpression expression) {
+    BasicDBObject letClause = new BasicDBObject();
+    letClause.putAll(
+        (Map<String, Object>)
+            expression.getLhs().accept(new MongoLetClauseSelectTypeExpressionVisitor()));
+    letClause.putAll(
+        (Map<String, Object>)
+            expression.getRhs().accept(new MongoLetClauseSelectTypeExpressionVisitor()));
+    return letClause;
+  }
+
+  private class MongoLetClauseSelectTypeExpressionVisitor implements SelectTypeExpressionVisitor {
+    @Override
+    public BasicDBObject visit(AggregateExpression expression) {
+      return new BasicDBObject();
+    }
+
+    @Override
+    public BasicDBObject visit(ConstantExpression expression) {
+      return new BasicDBObject();
```

Comment:
Is this deliberate? 
If this is not expected here, why not extend the `MongoSelectTypeExpressionParser`?

Reply:
yes, this is deliberate as this visitor is expected to go through SelectTypeExpressions and only return the fields that have to be declared in the let clause. So any other field other than AliasedIdentifierExpression has to be ignored but it should not throw an error in that case and for the case of AliasedIdentifierExpression it should return the field name that we need to add to the let clause.

Reply:
Got it. This may be a common use-case. So, can we create a common base class that returns default (empty map/BasicDBObject) for every implementation and then override it? 

That way
1. The default returning shared base-class can be re-used
2. Implementation here would look concise

Reply:
updated it